### PR TITLE
added support for tar.bz2, as well as fixed the previous decompressio…

### PR DESCRIPTION
**Pull request by nxpweilincao**
added support for tar.bz2 and it is able to decomporess normal bz2 file

Signed-off-by: Weilin Cao <weilin.cao@nxp.com>

### CODE REVIEW

File: libuuu/buffer.cpp [hunk 1]
```diff
@@ -767,14 +767,64 @@ int bz_async_load(string filename, shared_ptr<FileBuffer> p)
...
 
-int FSBz2::load(string backfile, string filename, shared_ptr<FileBuffer>p, bool async)
+bool is_pbzip2_file(string filename)
 {
-	if (filename != "*")
+	shared_ptr<FileBuffer> file=get_file_buffer(filename);
+	uint64_t filesize= file->size();
+	uint64_t readsize= (filesize< (1024*1024) )? filesize:(1024*1024); //read at most 1MB, because maximum block size is 900kb
+	int header_num=0;
+	char* ptr= (char*)file->data();
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: uint8_t, so need force type convert

File: libuuu/buffer.cpp [hunk 2]
```diff
@@ -767,14 +767,64 @@ int bz_async_load(string filename, shared_ptr<FileBuffer> p)
...
+	p->reserve(estimate_size);
+
+	int bzerr;
+	FILE *zStream = NULL;
+	zStream = fopen(compress_file.data(), "rb");
+
+	int nUnused=0;
+	unsigned char unused[BZ_MAX_UNUSED];
+	int nread;
+	unsigned char obuf[5000];
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: avoid allocate big array at stack

File: libuuu/buffer.cpp [hunk 3]
```diff
@@ -767,14 +767,64 @@ int bz_async_load(string filename, shared_ptr<FileBuffer> p)
...
+	unsigned char *unusedTmp;
+	uint64_t decompressed_size=0;
+	BZFILE* bzf = BZ2_bzReadOpen(&bzerr, zStream, 0, 0, unused, nUnused);
+	while(1)
+	{
+		nread = BZ2_bzRead(&bzerr, bzf, obuf, sizeof(obuf));
+		memcpy(p->data()+decompressed_size,obuf,nread);
+
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: decompressed_size may be bigger memory size

File: libuuu/buffer.cpp [hunk 4]
```diff
@@ -767,14 +767,64 @@ int bz_async_load(string filename, shared_ptr<FileBuffer> p)
...
+		decompressed_size+=nread;
+
+		if(bzerr==BZ_STREAM_END)
+			break;
+	}
+	p->resize(decompressed_size);
+
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: BZ2_bzclose

File: libuuu/buffer.cpp [hunk 5]
```diff
@@ -767,14 +767,84 @@ int bz_async_load(string filename, shared_ptr<FileBuffer> p)
...
+	bz_stream* strm= (bz_stream*)malloc(sizeof(bz_stream));
+	strm->bzalloc  = NULL;
+    strm->bzfree   = NULL;
+	strm->opaque   = NULL;
+
+	int ret;
+	ret = BZ2_bzDecompressInit (strm,0, 0 );
+	if (ret != BZ_OK)
 		return -1;
+	strm->next_in  = (char*)file;
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: filebuffer

File: libuuu/buffer.cpp [hunk 6]
```diff
@@ -767,14 +767,84 @@ int bz_async_load(string filename, shared_ptr<FileBuffer> p)
...
+
+	size=filebuffer->size();
+	file=filebuffer->data();
+
+	decompressed_file=p->data();
+	decompressed_size=0;
+	p->reserve(7*size);//the usual compressed ratio is about 18%, so 7*18% > 100%
+
+	bz_stream* strm= (bz_stream*)malloc(sizeof(bz_stream));
+	strm->bzalloc  = NULL;
```
Comment (nxpfrankli) [inline libuuu/buffer.cpp]: use stack to bz_stream

File: libuuu/buffer.cpp (hunk 5 above)
Comment (nxpweilincao) [reply libuuu/buffer.cpp]: i realized already used filebuffer for the FileBuffer ptr i created, so i rename "file" to "compressed_file", "size" to "compressed_size" for easier understanding.

**Pull request merged**